### ISSUE TITLE
Cooking Servo on Ubuntu14 (dependency issue)

### ISSUE THREAD

**Issue by ljgww** [servo/servo#21106]
I tried to build Servo using master branch.
Prerequisites as specified on https://github.com/servo/servo#setting-up-your-environment currently given as:

    sudo apt install git curl autoconf libx11-dev \
    libfreetype6-dev libgl1-mesa-dri libglib2.0-dev xorg-dev \
    gperf g++ build-essential cmake virtualenv python-pip \
    libssl1.0-dev libbz2-dev libosmesa6-dev libxmu6 libxmu-dev \
    libglu1-mesa-dev libgles2-mesa-dev libegl1-mesa-dev libdbus-1-dev \
    libharfbuzz-dev ccache clang

needed only few fixes in my case: libssl and virtualenv. After suggestions from github page, this worked for me: https://gist.github.com/ljgww/b585727a1b3f5f68dc97cd679f246ba0#file-build_servo_ubunty14_prereq-txt

Now, I try to explain the issue...

Typically, in my experience, when I compile something large I do not put too much expectations on the first compilation run. There is only few few 'large' binaries that worked out of box and compiled with least fuss. I am trying to learn Rust so having experience on compiling something large based on Rust is important to understand how whole system works in real life.

Usually a large package will fail because it misses some library that is not installed or not being built yet. Quite often after first compilation and fixing eventual missing dependencies that has to be  installed, the compilation and linking goes good. In some cases dependencies that are built are not always available, so sometimes it happens that after first compilation (and not fixing anything) then running second compilation things gets pass (Generally I tend to believe that in some code it is expected to see some library which needs to be compiled later, so after first run building code finds library ready). I am not sure what `make` thinks about hierarchical dependences but generally one will intuitively think that building systems would first make what they can and then make something what is dependent on something else (hierarchically)

I am quite impressed how Rust is talking relatively little while compiling (possibly suppressed warnings not to scare us) but liked clean output - Looked promising.

So If nothing is missing I expect that compiling and linking shall work with all relevant files.

Guided by such thought I enter:

     $ ./mach build --dev
 
After a lot of compiler lines (and quite some time (I went to sleep) - my pc is not example of speed) so after (what it seems not a long run) I read verdict:

    error: build failed
    Build FAILED in 1:26:12

Now I think OK lets run second run of compiling but checkout master branch so I am not building something else.

~/dev/rust/servo/servo$ git pull
Already up to date.
~/dev/rust/servo/servo$ git checkout master
Already on 'master'
Your branch is up to date with 'origin/master'.

Second run:

    ~/dev/rust/servo/servo$ ./mach build --dev
    Compiling mozjs_sys v0.51.2
    Compiling azure v0.29.0 (https://github.com/servo/rust-azure#1ca68f30)
    Compiling string_cache v0.7.3
    Compiling servo_atoms v0.0.1 (file:///home/lj/dev/rust/servo/servo/components/atoms)
    Compiling markup5ever v0.7.1
    Compiling cssparser v0.23.10
    error: failed to run custom build command for `mozjs_sys v0.51.2`
    process didn't exit successfully: `/home/lj/dev/rust/servo/servo/target/debug/build/mozjs_sys-ee5764afee0339d9/build-script-build` (exit code: 101)

    ... more text ...

    error: build failed
    Build FAILED in 0:01:25

For some unexplained reason I do third run...

    ~/dev/rust/servo/servo$ ./mach build --dev
    Compiling mozjs_sys v0.51.2
    Compiling selectors v0.19.0 (file:///home/lj/dev/rust/servo/servo/components/selectors)
    Compiling mime_guess v1.8.1
    Compiling script v0.0.1 (file:///home/lj/dev/rust/servo/servo/components/script)
    error: failed to run custom build command for `mozjs_sys v0.51.2`
    process didn't exit successfully: `/home/lj/dev/rust/servo/servo/target/debug/build/mozjs_sys-ee5764afee0339d9/build-script-build` (exit code: 101)

    .... more ....
    Build FAILED in 0:03:50

Then again

    ~/dev/rust/servo/servo$ ./mach build --dev
    Compiling mozjs_sys v0.51.2
    Compiling hyper v0.10.13
    error: failed to run custom build command for `mozjs_sys v0.51.2`

Did you notice? It becomes somewhat like pattern, every run will apparently compile something else

So...

    ~/dev/rust/servo/servo$ ./mach build --dev
    Compiling mozjs_sys v0.51.2
    Compiling url_serde v0.2.0
    Compiling ws v0.7.3
    error: failed to run custom build command for `mozjs_sys v0.51.2`
 
then

    ~/dev/rust/servo/servo$ ./mach build --dev
    Compiling mozjs_sys v0.51.2
    Compiling webrender v0.57.2 (https://github.com/servo/webrender#db26bc0e)
    error: failed to run custom build command for `mozjs_sys v0.51.2`

and so on

   Compiling mozjs_sys v0.51.2
   Compiling winit v0.11.3

next

   Compiling mozjs_sys v0.51.2
   Compiling string_cache v0.7.3
   Compiling selectors v0.19.0 (file:///home/lj/dev/rust/servo/servo/components/selectors)
   error: failed to run custom build command for `mozjs_sys v0.51.2`

next 

   Compiling mozjs_sys v0.51.2
   Compiling mime_guess v1.8.1
   Compiling hyper_serde v0.8.0
   Compiling hyper-openssl v0.2.6

I become determined to repeat it for as long as i get something compiling. Perhaps few more times (will include log for someone who wants to see what exactly happened. I finally got to only Compiling mozjs_sys v0.51.2 and filed another issue ( #21101 ).

SO this is what I find as an issue here. How this build system resolves hierarchical dependences? Are there such thing as hierarchical dependences? If there they are why it did not discover all dependences in one go and then re-do the higher levels? Is the building system broken? Or it is dependency of make (I have noticed number of c/c++ compilation steps going on underneath - so I would not blame Rust for which I believe Rust shall know about hierarchical dependences unless there is something I do not know (and I do not know much). Or perhaps is is some side effect of threaded runs ? (hint: 'warning: build failed, waiting for other jobs to finish...' message suggests to me that this may be the case). At any rate I find it peculiar and unusual and perhaps somewhere there is some possible bug.

And, here, included full cooking log (perhaps someone spots something else): https://gist.github.com/ljgww/c1ea58b40fcaa46a09a40edf490b5034#file-cooking_servo_ubuntu14-txt

**Comment by SimonSapin** [comment]
The reason you get different lines on each run is: to take advantage of multiple CPU cores, Cargo starts the compilation of multiple crates (packages) in parallel. When one of them, it stops starting new compilation jobs, but waits for the other running ones to finish. So as long as there is something else that does not depend on `mozjs_sys` that is still not built, restarting the build will make some more progress but still not finish.

```
error: unknown argument: '-fno-sized-deallocation', err: true
thread 'main' panicked at 'Should generate JSAPI bindings OK: ()', libcore/result.rs:945:5
```

This is likely bindgen failing because libclang is too old. It is possible that the README is not up to date, especially for Ubuntu 14.04 (which four years old now). The more reliable fix might be to update to a more recent Ubuntu. If that is not practical, try installing Clang 3.9 like this:

https://github.com/servo/servo/blob/e2fca1b228ff20d54b67e0d1fa502b694b5c290e/.travis.yml#L16-L19

**Comment by ljgww** [comment]
OK

    ~/dev/rust/servo/servo$ clang --version
    clang version 3.9.1-4ubuntu3~14.04.3 (tags/RELEASE_391/rc2)
    Target: x86_64-unknown-linux-gnu
    Thread model: posix
    InstalledDir: /usr/bin
    ~/dev/rust/servo/servo$ ./mach build --dev
     
and, failed again. https://gist.github.com/ljgww/a405854be742db26d820536f5b1f9165#file-clang3-9-servo_build-txt

I have mentioned in another posting (see related issue) that I of impression that for some reason clang is not being used

**Comment by SimonSapin** [comment]
CC @emilio regarding bindgen finding libclang

**Comment by ljgww** [comment]
btw #21101

**Comment by ljgww** [comment]
I had issues with my installation and managed to bring it back to life. Re-tried to build servo. Got a error message. Then decided to do: `git pull`
Got plenty of updates.
then rerun mach... now I get something else:

    ~/dev/rust/servo/servo$ ./mach build --dev
    "/home/lj/dev/rust/servo/servo/python/_virtualenv/bin/python2.7" "-m" "pip" "install" "-I" "-r" "/home/lj/dev/rust/servo/servo/tests/wpt/web-platform-tests/tools/wptrunner/requirements.txt" failed with error code 1:
    Output:
    Error:
    Could not import runpy module

This is puzzling because I seem not to find module RUNPY to install for Python2.7 but there is one for Python3. Now mach seems to rely on Python2.7 

Question is how to get runpy module?

Shall I submit another issue ticket?

**Comment by ljgww** [comment]
I have runpy but in default python setup...

    /usr/bin$ python
    Python 2.7.12 (default, Jan  9 2017, 12:16:27) 
    [GCC 4.8.4] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import runpy
    >>> exit()

Hence I have to clear servo's python virtualenv or trigger to rebuild it.

**Comment by dralley** [comment]
It should be noted that unless you're paying money to Canonical for extended support, Ubuntu 14.04 is no longer supported: https://www.omgubuntu.co.uk/2019/04/ubuntu-14-04-end-of-life
